Nested Facet nullable enabled error in generated code

Similar to #243 but includes Nested Facet.

In project nullable is enabled.

```cs
public class UserModel {
[Key] public int Id {get; set;}
public int SettingsId {get; set;}
public required UserSettingsModel Settings {get; set;} = new();
}

public class UserSettingsModel {
[Key] public int Id {get; set;}
// some settings as int, float, strings, ...
}

[Facet(typeof(UserSettingsModel), exclude: [nameof(UserSettingsModel.Id)])]
public class UserSettings {
   // ...
}

[Facet(typeof(UserModel), NestedFacets = [typeof(UserSettings)], PreserveRequiredProperties = true)]
public class User {
   public int ProcessedTicks => Settings.StopTick  - Settings.StartTick;
}
```

Maybe I am doing something wrong or dotnet ef should work other way.

- when remove `=new()` add `NullableProperties = fals`e I still get generated `User `with `public UserSettings? Settings;` so it's nullable 
- when remove `required` from Settings I get

>  Dereference of a possibly null reference.

- when stay like that there are errors on compile:

> error CS8601: Possible null reference assignment.
> error CS8618: Non-nullable property 'UserSettings' must contain a non-null value when exiting constructor. Consider adding the 'required' modifier or declaring the property as nullable.

on the lines of generated User class:
`this.User = __depth < 10 ? (__processed != null && ...`

--- a/docs/03_AttributeReference.md
+++ b/docs/03_AttributeReference.md
@@ -192,6 +192,62 @@ var mappedEmployee = employeeDto.ToSource();
 - Reference types (string, objects) remain reference types but are marked nullable
 - Disable `GenerateToSource` to avoid mapping issues from nullable to non-nullable types
 
+## Constructor Chaining
+
+The `ChainToParameterlessConstructor` parameter allows the generated constructor to chain to your user-defined parameterless constructor using `: this()`. This ensures any custom initialization logic in your constructor runs before property mapping.
+
+### Usage
+
+```csharp
+public class ModelType
+{
+    public int MaxValue { get; set; }
+    public string Name { get; set; } = string.Empty;
+}
+
+[Facet(typeof(ModelType), GenerateParameterlessConstructor = false, ChainToParameterlessConstructor = true)]
+public partial class MyDto
+{
+    public int Value { get; set; }
+    public bool Initialized { get; set; }
+
+    public MyDto()
+    {
+        // Custom initialization logic that runs before mapping
+        Value = 100;
+        Initialized = true;
+    }
+}
+
+// Usage
+var source = new ModelType { MaxValue = 42, Name = "Test" };
+var dto = new MyDto(source);
+// dto.Value == 100 (from parameterless constructor)
+// dto.Initialized == true (from parameterless constructor)
+// dto.MaxValue == 42 (from source mapping)
+// dto.Name == "Test" (from source mapping)
+```
+
+### Generated Code
+
+With `ChainToParameterlessConstructor = true`, the generated constructor chains to your parameterless constructor:
+
+```csharp
+public MyDto(ModelType source) : this()  // <-- Chains to your constructor
+{
+    this.MaxValue = source.MaxValue;
+    this.Name = source.Name;
+}
+```
+
+### When to Use
+
+- When you have initialization logic in your parameterless constructor that needs to run during mapping
+- When you need to set default values that aren't simply copied from the source
+- When you have computed or derived properties that need initial values
+
+**Note:** Set `GenerateParameterlessConstructor = false` to prevent the generator from creating its own parameterless constructor, which would conflict with yours.
+
 ## Attribute Copying
 
 The `CopyAttributes` parameter allows you to copy attributes from the source type's members to the generated facet members. This is particularly useful for preserving data validation attributes when creating DTOs for API models.

--- a/src/Facet/Generators/FacetGenerators/ExpressionBuilder.cs
+++ b/src/Facet/Generators/FacetGenerators/ExpressionBuilder.cs
@@ -192,7 +192,8 @@ internal static class ExpressionBuilder
                 return $"__depth < {maxDepth} && {sourceVariableName}.{sourcePropName} != null ? {collectionExpression} : null";
             }
 
-            return $"__depth < {maxDepth} ? {collectionExpression} : null";
+            // For non-nullable collections, use null-forgiving operator when depth is exceeded
+            return $"__depth < {maxDepth} ? {collectionExpression} : null!";
         }
         else
         {
@@ -248,7 +249,9 @@ internal static class ExpressionBuilder
             if (preserveReferences && useDepthParameter)
             {
                 // Check against __processed (not updatedProcessed) to detect if this exact object was already processed
-                return $"(__processed != null && __processed.Contains({sourceExpr}) ? null : {ctorCall})";
+                // For non-nullable types, use null-forgiving operator to avoid CS8601 compiler warnings
+                var nullFallback = isNullable ? "null" : "null!";
+                return $"(__processed != null && __processed.Contains({sourceExpr}) ? {nullFallback} : {ctorCall})";
             }
 
             return ctorCall;
@@ -264,7 +267,12 @@ internal static class ExpressionBuilder
                 return $"__depth < {maxDepth} && {sourceVariableName}.{sourcePropName} != null ? {constructorCall} : null";
             }
 
-            return $"__depth < {maxDepth} ? {constructorCall} : null";
+            // For non-nullable properties, use null-forgiving operator when depth is exceeded
+            // This is pragmatic because:
+            // 1. Circular references triggering this are rare
+            // 2. The depth limit is generous (default 10)
+            // 3. Required properties need a value, and null! is the best fallback
+            return $"__depth < {maxDepth} ? {constructorCall} : null!";
         }
         else
         {

--- a/test/Facet.Tests/TestModels/TestEntities.cs
+++ b/test/Facet.Tests/TestModels/TestEntities.cs
@@ -251,3 +251,41 @@ public partial class ChainedConstructorNoDepthDto
         Initialized = true;
     }
 }
+
+// When source has a required non-nullable nested property, 
+// the generated facet should respect that nullability
+public class UserModelWithRequiredSettings
+{
+    public int Id { get; set; }
+    public int SettingsId { get; set; }
+    public required UserSettingsModelForNested Settings { get; set; }
+}
+
+public class UserSettingsModelForNested
+{
+    public int Id { get; set; }
+    public int StartTick { get; set; }
+    public int StopTick { get; set; }
+}
+
+[Facet(typeof(UserSettingsModelForNested), nameof(UserSettingsModelForNested.Id))]
+public partial class UserSettingsFacet;
+
+// This should generate a non-nullable Settings property because:
+// 1. Source property is marked as 'required'
+// 2. Source property type has NotAnnotated nullable annotation (non-nullable)
+[Facet(typeof(UserModelWithRequiredSettings), PreserveRequiredProperties = true, NestedFacets = [typeof(UserSettingsFacet)])]
+public partial class UserWithRequiredSettingsFacet
+{
+    public int ProcessedTicks => Settings.StopTick - Settings.StartTick;
+}
+
+// Test without required - should be nullable for safety
+public class UserModelWithOptionalSettings
+{
+    public int Id { get; set; }
+    public UserSettingsModelForNested Settings { get; set; } = new();
+}
+
+[Facet(typeof(UserModelWithOptionalSettings), NestedFacets = [typeof(UserSettingsFacet)])]
+public partial class UserWithOptionalSettingsFacet;

--- a/test/Facet.Tests/UnitTests/Core/Facet/RequiredNestedFacetNullabilityTests.cs
+++ b/test/Facet.Tests/UnitTests/Core/Facet/RequiredNestedFacetNullabilityTests.cs
@@ -1,0 +1,161 @@
+using Facet.Tests.TestModels;
+
+namespace Facet.Tests.UnitTests.Core.Facet;
+
+/// <summary>
+/// When a source property is marked as 'required' and is non-nullable, 
+/// the generated facet should compile without nullable warnings.
+/// </summary>
+public class RequiredNestedFacetNullabilityTests
+{
+    [Fact]
+    public void RequiredNestedFacet_Property_ShouldBeNonNullable()
+    {
+        // Arrange & Act
+        var dtoType = typeof(UserWithRequiredSettingsFacet);
+        var settingsProperty = dtoType.GetProperty("Settings");
+
+        // Assert
+        settingsProperty.Should().NotBeNull();
+        // The property type should be the facet type (not nullable)
+        settingsProperty!.PropertyType.Should().Be(typeof(UserSettingsFacet));
+    }
+
+    [Fact]
+    public void RequiredNestedFacet_ShouldBeMarkedAsRequired()
+    {
+        // Arrange & Act
+        var dtoType = typeof(UserWithRequiredSettingsFacet);
+        var settingsProperty = dtoType.GetProperty("Settings");
+
+        // Assert
+        settingsProperty.Should().NotBeNull();
+        // Check if the property has the required modifier (via attribute in reflection)
+        var requiredAttribute = settingsProperty!.GetCustomAttributes(
+            typeof(System.Runtime.CompilerServices.RequiredMemberAttribute), true);
+        // Note: Required modifier appears on the type, not the property in reflection
+        // We verify by checking the type has RequiredMemberAttribute
+        var typeRequiredAttribute = dtoType.GetCustomAttributes(
+            typeof(System.Runtime.CompilerServices.RequiredMemberAttribute), true);
+        typeRequiredAttribute.Should().NotBeEmpty("Type should have RequiredMemberAttribute");
+    }
+
+    [Fact]
+    public void RequiredNestedFacet_Constructor_ShouldMapCorrectly()
+    {
+        // Arrange
+        var source = new UserModelWithRequiredSettings
+        {
+            Id = 1,
+            SettingsId = 100,
+            Settings = new UserSettingsModelForNested
+            {
+                Id = 100,
+                StartTick = 10,
+                StopTick = 50
+            }
+        };
+
+        // Act
+        var dto = new UserWithRequiredSettingsFacet(source);
+
+        // Assert
+        dto.Id.Should().Be(1);
+        dto.SettingsId.Should().Be(100);
+        dto.Settings.Should().NotBeNull();
+        dto.Settings.StartTick.Should().Be(10);
+        dto.Settings.StopTick.Should().Be(50);
+    }
+
+    [Fact]
+    public void RequiredNestedFacet_ComputedProperty_ShouldWork()
+    {
+        // Arrange
+        var source = new UserModelWithRequiredSettings
+        {
+            Id = 1,
+            SettingsId = 100,
+            Settings = new UserSettingsModelForNested
+            {
+                Id = 100,
+                StartTick = 10,
+                StopTick = 50
+            }
+        };
+
+        // Act
+        var dto = new UserWithRequiredSettingsFacet(source);
+
+        // Assert - The computed property should work without null reference issues
+        dto.ProcessedTicks.Should().Be(40); // 50 - 10
+    }
+
+    [Fact]
+    public void OptionalNestedFacet_Property_ShouldBeNullable()
+    {
+        // Arrange & Act
+        var dtoType = typeof(UserWithOptionalSettingsFacet);
+        var settingsProperty = dtoType.GetProperty("Settings");
+
+        // Assert
+        settingsProperty.Should().NotBeNull();
+        // For optional (non-required) nested facets, they should be treated as nullable
+        // The type will be the facet type (nullable reference type annotation is not visible via reflection)
+        settingsProperty!.PropertyType.Should().Be(typeof(UserSettingsFacet));
+    }
+
+    [Fact]
+    public void OptionalNestedFacet_Constructor_ShouldHandleValue()
+    {
+        // Arrange
+        var source = new UserModelWithOptionalSettings
+        {
+            Id = 2,
+            Settings = new UserSettingsModelForNested
+            {
+                Id = 200,
+                StartTick = 5,
+                StopTick = 25
+            }
+        };
+
+        // Act
+        var dto = new UserWithOptionalSettingsFacet(source);
+
+        // Assert
+        dto.Id.Should().Be(2);
+        dto.Settings.Should().NotBeNull();
+        dto.Settings!.StartTick.Should().Be(5);
+        dto.Settings.StopTick.Should().Be(25);
+    }
+
+    [Fact]
+    public void Projection_WithRequiredNestedFacet_ShouldWork()
+    {
+        // Arrange
+        var sources = new[]
+        {
+            new UserModelWithRequiredSettings
+            {
+                Id = 1,
+                SettingsId = 100,
+                Settings = new UserSettingsModelForNested
+                {
+                    Id = 100,
+                    StartTick = 0,
+                    StopTick = 100
+                }
+            }
+        }.AsQueryable();
+
+        // Act
+        var dtos = sources.Select(UserWithRequiredSettingsFacet.Projection).ToList();
+
+        // Assert
+        dtos.Should().HaveCount(1);
+        dtos[0].Id.Should().Be(1);
+        dtos[0].Settings.Should().NotBeNull();
+        dtos[0].Settings.StartTick.Should().Be(0);
+        dtos[0].Settings.StopTick.Should().Be(100);
+    }
+}

